simtools/production_configuration/calculate_statistical_errors_grid_point.py

```diff
@@ -0,0 +1,460 @@
+        sigma_energy = [np.std(d) if len(d) > 0 else np.nan for d in energy_deviation_by_bin]
+
+        # Calculate delta_energy as the mean deviation for each bin
+        delta_energy = [np.mean(d) if len(d) > 0 else np.nan for d in energy_deviation_by_bin]
+
+        # Combine sigma into a single measure
+        overall_uncertainty = np.nanmean(sigma_energy)
+
+        return overall_uncertainty, sigma_energy, delta_energy
+
+    def calculate_metrics(self):
+        """Calculate all defined metrics as specified in self.metrics and store results."""
+        print("self.metrics: ", self.metrics)
+        if "error_eff_area" in self.metrics:
+
+            self.error_eff_area = self.calculate_error_eff_area()
+            if self.error_eff_area:
+                validity_range = self.metrics.get("error_eff_area", {}).get("valid_range")
+                min_energy, max_energy = validity_range["value"][0] * u.Unit(
+                    validity_range["unit"]
+                ), validity_range["value"][1] * u.Unit(validity_range["unit"])
+
+                valid_errors = [
+                    error
+                    for energy, error in zip(
+                        self.data["bin_edges_low"], self.error_eff_area["relative_errors"]
+                    )
+                    if min_energy <= energy <= max_energy
+                ]
+                self.error_eff_area["max_error"] = max(valid_errors) if valid_errors else 0.0
+                ref_value = self.metrics.get("error_eff_area", {}).get("target_error")["value"]
+                _logger.info(
+                    f"Effective Area Error (max in validity range): "
+                    f"{self.error_eff_area['max_error'].value:.6f}, "
+                    f"Reference: {ref_value:.3f}"
+                )
+
+        if "error_energy_estimate_bdt_reg_tree" in self.metrics:
+            self.error_energy_estimate_bdt_reg_tree, self.sigma_energy, self.delta_energy = (
+                self.calculate_error_energy_estimate_bdt_reg_tree()
+            )
+            ref_value = self.metrics.get("error_energy_estimate_bdt_reg_tree", {}).get(
+                "target_error"
+            )["value"]
+            _logger.info(
+                f"Energy Estimate Error: {self.error_energy_estimate_bdt_reg_tree:.3f}, "
+                f"Reference: {ref_value:.3f}"
+            )
+        else:
+            raise ValueError("Invalid metric specified.")
+        self.metric_results = {
+            "error_eff_area": self.error_eff_area,
+            "error_energy_estimate_bdt_reg_tree": self.error_energy_estimate_bdt_reg_tree,
+        }
+        return self.metric_results
+
+    def calculate_max_error_for_effective_area(self):
+        """
+        Calculate the maximum relative error for effective area.
+
+        Returns
+        -------
+        max_error : float
+            Maximum relative error.
+        """
+        if "relative_errors" in self.metric_results["error_eff_area"]:
```

Comment:
Trying to understand a case where `relative_errors` is not in metric_results by `error_eff_area` is filled. If I understand it correctly, both variables are always filled in `calculate_metrics`?

Reply:
No currently this depends on the `production_simulation_config_metrics` file and what metrics are given there (i.e. which metric computation is required).